### PR TITLE
Improve components hooks docs

### LINKED ISSUE #12256: Improve Components Hooks docs
## How can Bevy's documentation be improved?

As someone who didn't followed closely #10756 I found a little confusing the different between `on_add` and `on_insert` when just reading the examples and docs.

Even tho I'm very eager to use component hooks (and we are closer to observers than relations) I think the docs and example could be improved a little bit:

1. `ComponentHooks::on_add` should state that spawning an entity will trigger `on_add` in the same way as `on_remove` says that despawning an entity counts as removing all of its components.

2. `ComponentHooks::on_insert` should state that mutating a component will not trigger `on_insert` like replacing the component will. I think this is important because in ECS, until now, there is no perceptual difference for the user between mutating a component or just replacing it. Now there is and users should be aware of that.
```rust
/// Will trigger on_insert hook
fn update_health(
    mut evts: EventReader<Damage>, q: Query<&Health, With<Player>>, mut commands: Commands,
) {
    for &Damage(entity, hit) in evts.read() {
        if let Ok(Health(current)) = q.get(entity) {
            let new_health = (current - hit).min(0.0);
            commands.entity(entity).insert(Health(new_health));
        }
    }
}
/// Won't trigger on_insert hook
fn update_health_mut(mut evts: EventReader<Damage>, mut q: Query<&mut Health, With<Player>>) {
    for &Damage(entity, hit) in evts.read() {
        if let Ok(mut current) = q.get_mut(entity) {
            let new_health = (current.0 - hit).min(0.0);
            current.0 = new_health;
        }
    }
}
```
3. `component_hooks.rs` example should avoid using "advanced" terms without an easy explanation for the user, like:
```
    // In order to register component hooks the component must:
    // - not belong to any created archetypes
    // - not already have a hook of that kind registered
```
In this case I think we should "translate" what `not belogin to any created archtypes` means, like "Can't be already inserted to any entity" or something like that.

4. `component_hooks.rs` example should (maybe?) tell users that `Events` should be prefered whenever applicable, since hooks can be more performance intensive or at least, more limited (I'm not sure about this one, tho)

If that make sense, I can make a PR for that.

### PR DESCRIPTION
**Pull request by afonsolage**
# Objective

- Closes #12256 

## Solution

- Improved component hooks docs.
- Improved component hooks example docs.

### CODE REVIEW

File: examples/ecs/component_hooks.rs [hunk 1]
```diff
@@ -1,4 +1,11 @@
-//! This examples illustrates the different ways you can employ component lifecycle hooks
+//! This example illustrates the different ways you can employ component lifecycle hooks.
+//!
+//! Whenever possible, prefer using Bevy's Events for reacting to component changes.
```
Comment (james7132) [inline examples/ecs/component_hooks.rs:3]:
```suggestion
//! Whenever possible, prefer using Bevy's change detection or Events for reacting to component changes.
```

File: examples/ecs/component_hooks.rs [hunk 2]
```diff
@@ -1,4 +1,11 @@
...
+//! Events generally offer better performance and more flexible integration into Bevy's systems.
+//! Hooks are useful in specific scenarios but have limitations (only one hook per component,
+//! less ergonomic than events).
+//!
+//! In the future, Bevy's Observer features might address some of these limitations.
```
Comment (james7132) [inline examples/ecs/component_hooks.rs:8]: Long term plans like these that may take a significant amount of time to materialize and get merged shouldn't be in a user-facing example.

File: examples/ecs/component_hooks.rs [hunk 3]
```diff
@@ -1,4 +1,11 @@
-//! This examples illustrates the different ways you can employ component lifecycle hooks
+//! This example illustrates the different ways you can employ component lifecycle hooks.
+//!
+//! Whenever possible, prefer using Bevy's Events for reacting to component changes.
+//! Events generally offer better performance and more flexible integration into Bevy's systems.
+//! Hooks are useful in specific scenarios but have limitations (only one hook per component,
+//! less ergonomic than events).
```
Comment (james7132) [inline examples/ecs/component_hooks.rs:6]: An example of when this feature might be useful (i.e. updating indexes, systems reliant on structural correctness like hierarchies).

**Review by pablo-lua — APPROVED** [review]

File: crates/bevy_ecs/src/component.rs [hunk 4]
```diff
@@ -221,17 +221,18 @@ pub struct ComponentHooks {
 
 impl ComponentHooks {
     /// Register a [`ComponentHook`] that will be run when this component is added to an entity.
-    /// An `on_add` hook will always be followed by `on_insert`.
+    /// An `on_add` hook will always be followed by `on_insert`. Spawning an entity counts as
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/component.rs:224]: What does `"will always be followed by `on_insert`" mean?

Comment (alice-i-cecile) [reply crates/bevy_ecs/src/component.rs:224]: I think this is just trying to say that these run before on_insert hooks?

File: examples/ecs/component_hooks.rs [hunk 5]
```diff
@@ -1,4 +1,17 @@
-//! This examples illustrates the different ways you can employ component lifecycle hooks
+//! This example illustrates the different ways you can employ component lifecycle hooks.
+//!
+//! Whenever possible, prefer using Bevy's change detection or Events for reacting to component changes.
+//! Events generally offer better performance and more flexible integration into Bevy's systems.
+//! Hooks are useful in specific scenarios but have limitations (only one hook per component,
```
Comment (alice-i-cecile) [inline examples/ecs/component_hooks.rs:5]:
```suggestion
//! Hooks are useful to enforce correctness but have limitations (only one hook per component,
```

File: crates/bevy_ecs/src/component.rs (hunk 4 above)
Comment (afonsolage) [reply crates/bevy_ecs/src/component.rs:224]: Yes, it means after `on_add` is ran, `on_insert` is always called also.

Comment (alice-i-cecile) [reply crates/bevy_ecs/src/component.rs:224]:
```suggestion
    /// An `on_add` hook will always run before `on_insert` hooks. Spawning an entity counts as
```

**Review by alice-i-cecile — APPROVED** [review]

**Pull request merged**